### ISSUE TITLE
[Feature Request]Allow for scanning of salesforce metadata

### ISSUE THREAD

**Issue by mgetzflex** [forcedotcom/sfdx-scanner#1237]
**Is your feature request related to an issue that you encountered with Salesforce Code Analyzer?**

Provide a way to do security and compliance checks against salesforce metadata, e.g. profiles, permissionsets, objects, etc. this can be done via PMD with custom xpath rules.

for example `sf scanner run -e pmd -pmdconfig ./mymetadatacompliancerules.xml  -t "**/*.profile,**/*.permissionset" -l metadata`  should run the custom rules against the profiles and permissionsets. 

**Describe the solution that you want:**

enable PMD to be run against the salesforce metadata xmls and provide away to set the pmd flag `--force-language xml` so that it recognizes the xmls for what they are

**Describe alternatives that you've considered:**

* Running PMD directly on the files
* Writing a custom tool using node/python etc.

**Additional Context:**

Ex rulefile that enforces that only profiles and permissionsets with "Admin" in the name can have the `ViewAllData` and `ModifyAllData` permissions

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="Compliance Permissions Checks"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description></description>
    <rule name="AdminViewAllData"
        language="xml"
        message="Only admin permissions should have ViewAllData"
        class="net.sourceforge.pmd.lang.xml.rule.DomXPathRule">
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
            //userPermissions[name/text()="ViewAllData"]/enabled[text()="true" and not(contains(lower-case(pmd:fileName()), 'admin'))]
            ]]></value>
            </property>
            <property name="defaultNsUri">
                <value>http://soap.sforce.com/2006/04/metadata</value>
            </property>
            <!-- Note: the property "version" is unsupported. -->
        </properties>
    </rule>
    <rule name="AdminModifyAllData"
        language="xml"
        message="Only admin permissions should have ModifyAllData"
        class="net.sourceforge.pmd.lang.xml.rule.DomXPathRule">
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
            //userPermissions[name/text()="ModifyAllData"]/enabled[text()="true" and not(contains(lower-case(pmd:fileName()), 'admin'))]
            ]]></value>
            </property>
            <property name="defaultNsUri">
                <value>http://soap.sforce.com/2006/04/metadata</value>
            </property>
            <!-- Note: the property "version" is unsupported. -->
        </properties>
    </rule>
</ruleset>
```

**Workaround:**

Run PMD directly

**Urgency:**

Nice to have, this reduces the duplication of tools as it means PMD is getting installed only once via the sf cli instead of twice.

**Comment by jfeingold35** [comment]
@mgetzflex, I'm not sure I understand what the actual ask is here, as far as net-new behavior goes.

The scanner in its current form allows you to enable PMD's built-in rules for XML by adding `xml` to PMD's `supportedLanguages` array in `~/.sfdx-scanner/Config.json`, or to add your own custom PMD rules for XML using `scanner rule add`. Also, while past releases included a default target pattern that prevented Salesforce metadata files from being targeted, that exclusion was removed in v3.17.0 (released October 3, 2023).

Are those existing features insufficient for your purposes here?

**Comment by mgetzflex** [comment]
@jfeingold35 As far as I'm aware there is no way to get the PMD embedded in the sfdx scanner to look at anything not ending in .xml and to treat it as xml. So despite the fact that profiles, permission sets, objects, etc. are all scannable xml... the scanner won't even consider loading PMD for that because they don't have the file extension `.xml`, `.cls`, or `.trigger`. PMD by default requires using the `--force-language` flag to make this work. You're thinking of `meta.xml` files, I'm talking the _actual metadata files themselves_ e.g. the `foo.profile`, `bar.permissionset`, `Account.object`, and `AccountBeforeUpdate.flow`.

My goal here is simple: be able to enforce best practice and compliance rules against the metadata, in particular profiles and permissionsets (until the former stops holding permissions in spring 2026). Can I use other tools for this? Yes. But this seems like a reasonable way to keeping everything together.

NOTE: the original post above was missing critical information because an html open comment was not closed before critical information. I've edited it to close that appropriately and make the relevant info visible.

**Comment by jfeingold35** [comment]
@mgetzflex , I understand. Thanks for clarifying. The team needs to have a deeper discussion about whether this is something we can realistically add with the resources currently available to us. I should be able to provide a decision on that front in the next few days.

In the meantime, thanks for bringing our attention to the missing `-->` tag in the issue template. That has been remedied.

**Comment by jfeingold35** [comment]
@mgetzflex , the team has talked it over, and we've decided against supporting this.
The metadata format (`whatever.object`) is quite old, and discouraged in favor of the less-old source format (`whatever.object-meta.xml`), which `sfdx-scanner` already supports.

**Comment by vt89** [comment]
@jfeingold35 Any change that the Salesforce Code Analyzer will support files available in Metadata Format such as `.object`, `.permissionset`?
It is a blocker for projects still using the Metadata Format instead of the Source Format since it is not possible to implement custom PMD rules for these metadata.
